iroh/src/baomap/mem.rs

```diff
@@ -0,0 +1,497 @@
+//! A full in memory database for iroh-bytes
+//!
+//! Main entry point is [Store].
+use std::collections::BTreeMap;
+use std::io;
+use std::num::TryFromIntError;
+use std::ops::DerefMut;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::sync::RwLock;
+
+use bao_tree::blake3;
+use bao_tree::io::fsm::Outboard;
+use bao_tree::io::outboard::PreOrderOutboard;
+use bao_tree::io::outboard_size;
+use bao_tree::BaoTree;
+use bao_tree::ByteNum;
+use bytes::Bytes;
+use bytes::BytesMut;
+use derive_more::From;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use iroh_bytes::baomap;
+use iroh_bytes::baomap::ExportMode;
+use iroh_bytes::baomap::ImportMode;
+use iroh_bytes::baomap::ImportProgress;
+use iroh_bytes::baomap::PartialMap;
+use iroh_bytes::baomap::PartialMapEntry;
+use iroh_bytes::baomap::ValidateProgress;
+use iroh_bytes::baomap::{Map, MapEntry, ReadonlyStore};
+use iroh_bytes::util::progress::IdGenerator;
+use iroh_bytes::util::progress::ProgressSender;
+use iroh_bytes::{Hash, IROH_BLOCK_SIZE};
+use iroh_io::AsyncSliceReader;
+use iroh_io::AsyncSliceWriter;
+use range_collections::RangeSet2;
+use tokio::sync::mpsc;
+
+use super::flatten_to_io;
+
+/// A mutable file like object that can be used for partial entries.
+#[derive(Debug, Clone, Default)]
+#[repr(transparent)]
+pub struct MutableMemFile(Arc<RwLock<BytesMut>>);
+
+impl MutableMemFile {
+    /// Create a new empty file
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self(Arc::new(RwLock::new(BytesMut::with_capacity(capacity))))
+    }
+
+    /// Freeze the data, returning the content
+    ///
+    /// Note that this will clear other references to the data.
```

Comment:
why clear those, instead of freezing them?

Reply:
You are taking the BytesMut out of the Arc<Mutex<...>> using std::mem::swap. Freeze takes a BytesMut, not a &mut BytesMut, so this is the only way I could find to make this zero copy.